File: src/main/java/com/apps/PortfolioApp/Service/SkillService.java
package com.apps.PortfolioApp.Service;

import com.apps.PortfolioApp.DTO.SkillDTO;
import com.apps.PortfolioApp.ExceptionHandler.DuplicateSkillException;
import com.apps.PortfolioApp.ExceptionHandler.SkillNotFoundException;
import com.apps.PortfolioApp.Model.Entity.Skill;
import com.apps.PortfolioApp.Repository.SkillRepository;
import org.springframework.stereotype.Service;
import java.util.List;

@Service
public class SkillService {

    // step1 ----> Create an object from SkillRepository class :
    private final SkillRepository skillRepository;

    // step2 ----> Create a con :
    public SkillService(SkillRepository skillRepository) {
        this.skillRepository = skillRepository;
    }

    // step3 ----> here I will Create a function to Insert to db :
    public SkillDTO insertSkill(SkillDTO skillDTO) {
        if (skillRepository.findByName(skillDTO.getName()).isPresent()) {
            throw new DuplicateSkillException("Skill already exists with name:" + skillDTO.getName());
        }
        Skill skill = SkillDTO.fromDTOToEntity(skillDTO);
        return SkillDTO.fromEntityToDTO(skillRepository.save(skill));
    }

    // step4 ----> here I will Create a function to get all Skills from db :
    public List<SkillDTO> getAllSkills() {
        if (skillRepository.findAll().isEmpty()) {
            throw new SkillNotFoundException("You don't have any skill");
        } else {
            List<Skill> skills = skillRepository.findAll();
            return skills.stream().map(SkillDTO::fromEntityToDTO).toList();
        }
    }

    // step5 ----> here I will Create a function to get a skill by id :
    public SkillDTO getSkillByID(int id) {
        Skill skill = skillRepository.findById(id).orElseThrow(
                () -> new SkillNotFoundException("Skill not found with id: " + id));
        return SkillDTO.fromEntityToDTO(skill);
    }

    // step6 ----> here I will Create a function update a skill from db :
    public SkillDTO updateSkill(String name, SkillDTO skillDTO) {
        Skill skill = skillRepository.findByName(name).orElseThrow(
                () -> new SkillNotFoundException("Skill not found with name: " + name));
        skill.setName(skillDTO.getName());
        skill.setLevel(skillDTO.getLevel());
        skill.setIcon(skillDTO.getIcon());
        return SkillDTO.fromEntityToDTO(skillRepository.save(skill));
    }

    // step7 ----> here I will Create a function delete a skill from db :
    public void deleteSkill(int id) {
        Skill skill = skillRepository.findById(id).orElseThrow(
                () -> new SkillNotFoundException("Skill not found with id: " + id));
        skillRepository.delete(skill);
    }

    // step8 ----> here I will Create a function to find by Skill name :
    public SkillDTO getByName(String name) {
        Skill skill = skillRepository.findByName(name).orElseThrow(
                () -> new SkillNotFoundException("Skill not found with skill name: " + name));
        return SkillDTO.fromEntityToDTO(skill);
    }

}

File: lib/services/skill_service.dart
import 'dart:convert';

import 'package:http/http.dart' as http;
import 'package:protfolio_app/exceptions/duplicate_skill_exception.dart';
import 'package:protfolio_app/exceptions/not_found_exception.dart';
import '../utils/constants.dart';
import 'package:protfolio_app/Models/skill.dart';

class SkillService {
  //step1 ----> here I will Create function to get all skills :
  Future<List<Skill>> getAllSkills() async {
    final response = await http.get(Uri.parse("${Constants.baseUrl}/skills"));

    if (response.statusCode == 200) {
      final List<dynamic> data = jsonDecode(response.body);

      return data.map((skill) => Skill.fromJson(skill)).toList();
    }

    if (response.statusCode == 404) {
      throw NotFoundException("Skill not found");
    }

    throw Exception("Failed");
  }

  //step2 ----> here I will Create a function to insert skill :
  Future<Skill> addSkill(Skill skill) async {
    final response = await http.post(
      Uri.parse("${Constants.baseUrl}/skills"),
      headers: {"Content-Type": "application/json"},
      body: jsonEncode({
        "name": skill.name,
        "level": skill.level,
        "icon": skill.icon,
      }),
    );

    if (response.statusCode == 200 || response.statusCode == 201) {
      return Skill.fromJson(jsonDecode(response.body));
    }

    if (response.statusCode == 409) {
      throw DuplicateSkillException("Skill already exists");
    }

    throw Exception("Failed to add skill");
  }

  //step3 ----> here I will Create a function to update skill :
  Future<Skill> updateSkill(String oldName, Skill skill) async {
    final response = await http.put(
      Uri.parse("${Constants.baseUrl}/skills/$oldName"),
      headers: {"Content-Type": "application/json"},
      body: jsonEncode({
        "name": skill.name,
        "level": skill.level,
        "icon": skill.icon,
      }),
    );

    if (response.statusCode == 200) {
      return Skill.fromJson(jsonDecode(response.body));
    }

    if (response.statusCode == 404) {
      throw NotFoundException("Skill not found");
    }

    print(response.body);
    throw Exception("Failed to update skill");
  }

  //step4 ----> here I will Create a function to delete skill :
  Future<void> deleteSkill(int id) async {
    final response = await http.delete(
      Uri.parse("${Constants.baseUrl}/skills/$id"),
    );

    if (response.statusCode == 200 || response.statusCode == 204) {
      return;
    }

    if (response.statusCode == 404) {
      throw NotFoundException("Skill not found");
    }

    throw Exception("Failed to delete skill");
  }
}
